helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/algorithm/ConstraintBasedAlgorithm.java

```diff
@@ -0,0 +1,109 @@
+                    // adjust the node's capacity and fault zone information
+                    node.assign(replica);
+                    clusterContext.addPartitionToFaultZone(node.getFaultZone(), resource, replica.getPartitionName());
+                });
+                // stop immediately if any replica cannot find best assignable node
+                if (currentAssignment.hasAnyFailure()) {
+                    LOG.info("The current assignment encountered failure, clean up and rolling back to origin cluster context");
+                    // roll-back previous assignments
+                    currentAssignment.getCurrentAssignment().entrySet()
+                            .forEach(replicaToNodes ->
+                                    replicaToNodes.getValue().forEach(node -> node.release(replicaToNodes.getKey()))
+                            );
+                    clusterContext.resetFaultZoneMap();
+                    return currentAssignment;
+                }
+            }
+        }
+
+        return currentAssignment;
+    }
+
+    private Optional<AssignableNode> getNodeWithHighestPoints(
+            AssignableReplica replica, List<AssignableNode> assignableNodes, ClusterContext clusterContext, OptimalAssignment currentAssignment) {
+        // Only record the first hard constraint failure reason of the node
+        Map<String, HardConstraint> failedHardConstraintOfNode = new HashMap<>();
+        List<AssignableNode> candidateNodes = assignableNodes.stream()
+                .filter(candidate -> _hardConstraints.stream()
+                        .allMatch(constraint -> {
+                            boolean isValid = constraint.isAssignmentValid(candidate, replica, clusterContext);
+                            if (!isValid) {
+                                failedHardConstraintOfNode.put(candidate.getInstanceName(), constraint);
+                            }
+                            return isValid;
+                        }))
+                .collect(Collectors.toList());
+        if (candidateNodes.isEmpty()) {
+            LOG.error("Unable to find any available candidate nodes for partition {}; Fail reasons: {}",
+                    replica.getPartitionName(), failedHardConstraintOfNode);
+            //TODO: re-format the error message, the current error message is string format of a node -> HardConstraint map
+            currentAssignment.reportFailure(replica.getPartitionName(), failedHardConstraintOfNode.toString());
+            return Optional.empty();
+        }
+
+        Function<AssignableNode, Double> calculatePoints =
+                (candidateNode) -> _softConstraints.stream()
+                        .map(softConstraint -> softConstraint.assignmentScore(candidateNode, replica, clusterContext))
+                        .mapToDouble(score -> score)
+                        .sum();
+
+        AssignableNode bestNode = candidateNodes.stream()
```

Comment:
You can directly stream based on the previous sum, right?
calculatePoints is not quite necessary.

Reply:
`calculatePoints` is a function expression, a concise way of expressing method. It's not actually computing anything without line 104

Reply:
Got it.